rfcs/2021-05-05-7351-framing-and-codecs.md

```diff
@@ -0,0 +1,183 @@
+are implemented.
+
+Framing takes place before an event is created since the message boundary
+required to form an event is not known yet. Therefore the source needs to be
+aware of framing and call the framer to determine message boundaries before
+creating an event.
+
+To make the source aware of framing, the `Framer` needs to be passed to the
+`SourceContext`, so that the source implementation can call it:
+
+```rust
+struct SourceContext {
+    // … existing fields
+    framing: Framer,
+}
+```
+
+In the context of Vector, decoders and encoders implement functions with the
+signature `fn decode(value: Value) -> Result<Value>` /
+`fn encode(value: Value) -> Result<Value>` where e.g. a `json` encoder can
+convert a `Value::Bytes` to `Value::Map`.
+
+For the implementation of a framer, we can defer to Tokio's
+[`FramedRead`](https://docs.rs/tokio-util/0.6.6/tokio_util/codec/struct.FramedRead.html)
+trait which does exactly what we want here.
+
+## Doc-level Proposal
+
+// TODO.
+
+<!--
+- Optional. Only do this if your change is public facing.
+- Demonstrate how your change will look in the form of Vector's public docs.
+-->
+
+## Rationale
+
+One prime example where a source's implementation may be reused with a different
+codec is the `syslog` source (see
+[#7046](https://github.com/timberio/vector/pull/7046)), or the upcoming `syslog`
+sink in [#7106](https://github.com/timberio/vector/issues/7106). Instead of
+re-implementing socket-based connection handling, the `syslog` components could
+be replaced by the `socket` counterparts combined with `octet-framing`. This
+reduces a possible source of bugs and inconsistencies and therefore leads to
+less maintenance burden.
+
+Introducing codecs may also shrink unnecessary noise in config files by removing
+transform steps / input indirections, when basic transforms were used that are
+only concerned with encoding formats.
+
+<!--
+- Why is this change worth it?
+- What is the impact of not doing this?
+- How does this position us for success in the future?
+-->
+
+## Prior Art
+
+// TODO.
+
+- Tokio Codecs.
+
+<!--
+- List prior art, the good and bad.
+- Why can't we simply use or copy them?
+-->
+
+## Drawbacks
+
+// TODO.
+
+- Is this the right abstraction, is it possible to cleanly separate at these
+  boundaries for all our current and future use cases?
+- For the components that we expect benefit most from reusability enabled by
+  separating, will they always only differ by encoding in the future?
+- Does it hurt discoverability (e.g. predefined `syslog` source vs `socket` +
+  `octet-framing`)?
+
+<!--
+- Why should we not do this?
+- What kind on ongoing burden does this place on the team?
+-->
+
+## Alternatives
+
+// TODO.
+
+<!--
+- What other approaches have been considered and why did you not choose them?
+- How about not doing this at all?
+-->
+
+## Outstanding Questions
+
+Is there a realistic chance that multiple framing / codec options might need be
+applied / composed?
```

Comment:
If we are also going to extract compression codecs into these abstractions, then I think yes.

Reply:
Compression codecs are likely not to apply to individual messages, but whole requests at the sink level. We could potentially expose this via the same config, but it seems like the implementation would be quite different.

Reply:
I'm not sure how common that is, but I could imagine `gzip`'d bundles of line-delimited JSON messages that are sent via UDP.

In any case, if we decide to inject the codecs at the topology level and the codecs accept `Event` as input, the chaining would be supported more or less naturally.

Chaining framing would also work, as they operate on `Vec<Bytes> -> Vec<Bytes>`. One example could be first delimiting messages by line-ending and then by any custom character, e.g. whitespace.